### ISSUE TITLE
Maintain an optional CRC for each file.

### ISSUE THREAD

**Issue by AudriusButkevicius** [syncthing/syncthing#1711]
To detect changes without mtime changing ?

**Comment by calmh** [comment]
Why, just because it's faster than SHA256?

**Comment by AudriusButkevicius** [comment]
And much cheaper.

**Comment by Zillode** [comment]
Also useful for https://github.com/syncthing/syncthing/pull/1852
We need to check the CRC before overwriting a file during a pull.

**Comment by calmh** [comment]
Can we not just check the mtime, and if it differs we anyway need to do the full scan?

**Comment by jorangreef** [comment]
It's not enough to check the mtime, since many filesystems have a mtime resolution of 1 second it's possible that a write could be missed. Unless there was a delay before hashing and storing that file in the index the first time.

And even hashing the file using CRC or SHA256 would not prevent updates from being lost. For example:

1. An application has a file descriptor open to the file.
2. Syncthing updates the file to a newer version received from another peer using an atomic replace.
3. The application writes to the file descriptor 2 minutes later.

Dropbox actually handles the above edge case. Whenever it updates a file, it moves the old version to a quarantine directory within `.dropbox`. It then monitors this directory for changes, and creates an update conflict if the quarantined file changes.

**Comment by AudriusButkevicius** [comment]
Fds are associated with an inode not filenames, so writes after the file is replaced are basically written to a file which is no longer on the filesystem.

**Comment by jorangreef** [comment]
Here's another example:

1. An application on peer 1 has a file descriptor open to a file.
2. The file is deleted on peer 2.
3. Syncthing deletes the file on peer 1.
4. The application on peer 1 now updates the file via the file descriptor.
5. The file is restored on peer 2.
6. Syncthing restores the original version on peer 1.
7. The update made at step 4 is now lost.

**Comment by AudriusButkevicius** [comment]
Yes, but there is no way to protect against this, all applications that I am aware of would loose data if they had the rug pulled from underneath their feet.

Tracking open FD's is not an option, as that does work well cross-platform and requires administrator privileges on some.

**Comment by jorangreef** [comment]
There is actually a very simple solution to catch these lost updates and you don't need to track open FDs (that's prone to race conditions anyway).

Dropbox is already doing it. They use a quarantine directory (see my [first comment](https://github.com/syncthing/syncthing/issues/1711#issuecomment-128273270)). You can test out those scenarios for yourself with Dropbox. It handles them fine.

**Comment by jorangreef** [comment]
It's also cheaper then trying to rehash the file before you update it (this is also prone to race conditions).

**Comment by AudriusButkevicius** [comment]
You say it monitors the directory for changes, what does that mean?

**Comment by Zillode** [comment]
when we are done pulling, we should move the original file and do something like:
```
os.Rename(filename, oldfile)
defer func() {
  time.Sleep(5*time.Minute)
  if crc(oldfile) != crc_in_db_before_update) { create_conflict() }
  else { delete(oldfile) }
}
```

**Comment by calmh** [comment]
How so?

**Comment by AudriusButkevicius** [comment]
Don't we have now check mtime in finish?

**Comment by calmh** [comment]
The first couple of comments on this thread are confusing. But the thing where some program has an open fd on a file that we replace, and later decides to write to that fd, is still an issue afaik. Although probably not a very common one.
